module Nextbillionai
  module Errors
    class Error < StandardError
      attr_accessor cause: StandardError?
    end

    class ConversionError < Nextbillionai::Errors::Error
      def cause: -> StandardError?

      def initialize: (
        on: Class,
        method: Symbol,
        target: top,
        value: top,
        ?cause: StandardError?
      ) -> void
    end

    class APIError < Nextbillionai::Errors::Error
      attr_accessor url: URI::Generic

      attr_accessor status: Integer?

      attr_accessor body: top?

      def initialize: (
        url: URI::Generic,
        ?status: Integer?,
        ?body: Object?,
        ?request: nil,
        ?response: nil,
        ?message: String?
      ) -> void
    end

    class APIConnectionError < Nextbillionai::Errors::APIError
      def initialize: (
        url: URI::Generic,
        ?status: nil,
        ?body: nil,
        ?request: nil,
        ?response: nil,
        ?message: String?
      ) -> void
    end

    class APITimeoutError < Nextbillionai::Errors::APIConnectionError
      def initialize: (
        url: URI::Generic,
        ?status: nil,
        ?body: nil,
        ?request: nil,
        ?response: nil,
        ?message: String?
      ) -> void
    end

    class APIStatusError < Nextbillionai::Errors::APIError
      def self.for: (
        url: URI::Generic,
        status: Integer,
        body: Object?,
        request: nil,
        response: nil,
        ?message: String?
      ) -> instance

      def initialize: (
        url: URI::Generic,
        status: Integer,
        body: Object?,
        request: nil,
        response: nil,
        ?message: String?
      ) -> void
    end

    class BadRequestError < Nextbillionai::Errors::APIStatusError
      HTTP_STATUS: 400
    end

    class AuthenticationError < Nextbillionai::Errors::APIStatusError
      HTTP_STATUS: 401
    end

    class PermissionDeniedError < Nextbillionai::Errors::APIStatusError
      HTTP_STATUS: 403
    end

    class NotFoundError < Nextbillionai::Errors::APIStatusError
      HTTP_STATUS: 404
    end

    class ConflictError < Nextbillionai::Errors::APIStatusError
      HTTP_STATUS: 409
    end

    class UnprocessableEntityError < Nextbillionai::Errors::APIStatusError
      HTTP_STATUS: 422
    end

    class RateLimitError < Nextbillionai::Errors::APIStatusError
      HTTP_STATUS: 429
    end

    class InternalServerError < Nextbillionai::Errors::APIStatusError
      HTTP_STATUS: Range[Integer]
    end
  end
end
